File: acceleration.f90
MODULE PERIOD_ACCELERATION_MDL
    INTERFACE
         SUBROUTINE ACCELERATION_OF_PERIODS( ALPHAS, &
                                    CURRENT_PERIODS, &
                                    INTERNAL_STRESS, &
                                    EXTERNAL_STRESS, &
                               PERIOD_ACCELERATIONS   )
             IMPLICIT NONE
             REAL*8, INTENT(IN) :: ALPHAS(3),            &
                                   CURRENT_PERIODS(3,3), &
                                   INTERNAL_STRESS(3,3), &
                                   EXTERNAL_STRESS(3,3)
             REAL*8, INTENT(OUT):: PERIOD_ACCELERATIONS(3,3)
         END SUBROUTINE ACCELERATION_OF_PERIODS
    END INTERFACE
END MODULE PERIOD_ACCELERATION_MDL




         SUBROUTINE ACCELERATION_OF_PERIODS( ALPHAS, &
                                    CURRENT_PERIODS, &
                                    INTERNAL_STRESS, &
                                    EXTERNAL_STRESS, &
                               PERIOD_ACCELERATIONS   )
             IMPLICIT NONE
             REAL*8, INTENT(IN) :: ALPHAS(3),            &
                                   CURRENT_PERIODS(3,3), &
                                   INTERNAL_STRESS(3,3), &
                                   EXTERNAL_STRESS(3,3)
             REAL*8, INTENT(OUT):: PERIOD_ACCELERATIONS(3,3)
             REAL*8             :: CELL_SURFACES(3,3)
             REAL*8             :: CELL_VOLUME
             INTEGER            :: I, J, K, L, M, N

             DO I = 1, 3
                J = MOD(I, 3) + 1
                K = MOD(J, 3) + 1
                DO L = 1, 3
                   M = MOD(L, 3) + 1
                   N = MOD(M, 3) + 1
                   CELL_SURFACES(L,I) =                              &
                         CURRENT_PERIODS(M,J) * CURRENT_PERIODS(N,K) &
                       - CURRENT_PERIODS(N,J) * CURRENT_PERIODS(M,K)
                END DO
             END DO

             CELL_VOLUME = CELL_SURFACES(1,1) * CURRENT_PERIODS(1,1) &
                         + CELL_SURFACES(2,1) * CURRENT_PERIODS(2,1) &
                         + CELL_SURFACES(3,1) * CURRENT_PERIODS(3,1)

             IF(CELL_VOLUME <= 1.0D-10) THEN
                PRINT*, "Sorry, too small or negative cell volume:"
                PRINT*, CELL_VOLUME
                STOP
             END IF

             DO I = 1, 3
                DO J = 1, 3
                   PERIOD_ACCELERATIONS(J, I) = 0.0D0
                   DO K = 1, 3
                      PERIOD_ACCELERATIONS(J, I) = PERIOD_ACCELERATIONS(J, I) + &
                              (INTERNAL_STRESS(J, K) + EXTERNAL_STRESS(J, K)) * &
                                                          CELL_SURFACES(K, I)
                   END DO
                   PERIOD_ACCELERATIONS(J, I) = PERIOD_ACCELERATIONS(J, I) &
                                              / ALPHAS(I)
                END DO
             END DO

             RETURN
         END SUBROUTINE ACCELERATION_OF_PERIODS

File: acceleration.c
#include <stdio.h>
#include <stdlib.h>


void Acceleration_of_PERIODS( double * alphas,
                              double * current_periods,
                              double * internal_stress,
                              double * external_stress,
                              double * period_accelerations )
{    double cell_surfaces[3][3], cell_volume;
     int               i, j, i3j, k, l, m, n;

     for(i=0; i<3; i++)
        {j = (i+1)%3;
         k = (j+1)%3;
         for(l=0; l<3; l++)
            {m = (l+1)%3;
             n = (m+1)%3;
             cell_surfaces[i][l] =
                 current_periods[j*3+m] * current_periods[k*3+n]
               - current_periods[j*3+n] * current_periods[k*3+m];
            }
        }

     cell_volume = cell_surfaces[0][0] *  current_periods[0]
                 + cell_surfaces[0][1] *  current_periods[1]
                 + cell_surfaces[0][2] *  current_periods[2];

     if(cell_volume < 1.0e-10)
       {printf(" Sorry, too small or negative cell volume: ");
        printf(" %f .\n", cell_volume);
        exit(0);
       }

     for(i=0; i<3; i++)
        {for(j=0; j<3; j++)
            {i3j = i * 3 + j;
             period_accelerations[i3j] = 0.0000000000000000000e0;
             for(k=0; k<3; k++)
                {period_accelerations[i3j]  +=
                    (internal_stress[j*3+k] + external_stress[j*3+k]) *
                                                 cell_surfaces[i][k]  ;
                }
             period_accelerations[i3j] = period_accelerations[i3j]
                                       / alphas[i]               ;
            }
        }
}

